From 6e5fe6311e0e9b6f14f4909eff035060a0314234 Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Wed, 10 Sep 2014 18:28:27 -0400
Subject: [ethernet] ixgbe: modify behavior on receiving a HW ECC error

Message-id: <1410373750-11341-8-git-send-email-jogreene@redhat.com>
Patchwork-id: 93561
O-Subject: [RHEL7.1 PATCH 07/50] ixgbe: modify behavior on receiving a HW ECC error.
Bugzilla: 1091123
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

BZ: 1091123
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7919168
Tested: In beaker. See BZ for details

Currently when we noticed a HW ECC error we would request the use reload
the driver to force a reset of the part.  This was done due to the mistaken
believe that a normal reset would not be sufficient.  Well it turns out it
would be so now we just schedule a reset upon seeing the ECC.

Signed-off-by: Don Skidmore <donald.c.skidmore@intel.com>
Tested-by: Phil Schmitt <phillip.j.schmitt@intel.com>
Signed-off-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit d773ce2de1c670e0d259870a2ea8fd9f60ab98cd)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c b/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
index 3e3a9b9..45ce567 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe_main.c
@@ -2593,9 +2593,12 @@ static irqreturn_t ixgbe_msix_other(int irq, void *data)
  switch (hw->mac.type) {
  case ixgbe_mac_82599EB:
  case ixgbe_mac_X540:
-  if (eicr & IXGBE_EICR_ECC)
-   e_info(link, "Received unrecoverable ECC Err, please "
-          "reboot\n");
+  if (eicr & IXGBE_EICR_ECC) {
+   e_info(link, "Received ECC Err, initiating reset\n");
+   adapter->flags2 |= IXGBE_FLAG2_RESET_REQUESTED;
+   ixgbe_service_event_schedule(adapter);
+   IXGBE_WRITE_REG(hw, IXGBE_EICR, IXGBE_EICR_ECC);
+  }
   /* Handle Flow Director Full threshold interrupt */
   if (eicr & IXGBE_EICR_FLOW_DIR) {
    int reinit_count = 0;
@@ -2809,9 +2812,12 @@ static irqreturn_t ixgbe_intr(int irq, void *data)
   ixgbe_check_sfp_event(adapter, eicr);
   /* Fall through */
  case ixgbe_mac_X540:
-  if (eicr & IXGBE_EICR_ECC)
-   e_info(link, "Received unrecoverable ECC err, please "
-         "reboot\n");
+  if (eicr & IXGBE_EICR_ECC) {
+   e_info(link, "Received ECC Err, initiating reset\n");
+   adapter->flags2 |= IXGBE_FLAG2_RESET_REQUESTED;
+   ixgbe_service_event_schedule(adapter);
+   IXGBE_WRITE_REG(hw, IXGBE_EICR, IXGBE_EICR_ECC);
+  }
   ixgbe_check_overtemp_event(adapter, eicr);
   break;
  default:
-- 
1.7.1